File: Cpp Course Assignments/1 - RPN Calculator in C, Bash, and Python/rpnc.c
/*
 * Annie Shen (ashen7)
 * CMPS 109 Harrison - Spring 2018
 * Lab 1 - C Reverse Polish Notation (RPN) Calculator
 * Due Apr 8th, 2018
 *
 * The entire RPN expression is actually passed as a SINGLE argument!
 * Returns -1 on error, 0 otherwise.
 */

/* Resources:
 * String Parsing: https://www.tutorialspoint.com/c_standard_library/c_function_strtok.htm
 * Converting char to int using atoi: https://stackoverflow.com/questions/628761/convert-a-character-digit-to-the-corresponding-integer-in-c
 */


#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include <math.h>

int main(int argc, char *argv[]) {
  const char s[1] = " ";
  char *token;
  token = strtok(argv[1], s); // Get first token
  int operand[16];
  int j = -1;
  int num1 = 0;
  int num2 = 0;
  int result = 0;
  int dbool = 1;

  while (token != NULL) {
    //printf("here\n");
    /* Loop through the token itself due to isdigit limitations */
    for (int i = 0; i < strlen(token); i++) {
      //printf("token:%c ",token[i]);
      if (isdigit(token[i]) == 0) {
        dbool = 0;
      }
    }

    /* If dbool == 1 -> the current token is a number! Else, an operator. */
    if (dbool == 1) {
      operand[++j] = atoi(token);

    /* Must be a operator */
    } else {
      num2 = operand[j];
      num1 = operand[j-1];

      if (strcmp(token, "+") == 0) {
        result = num1 + num2;
      } else if (strcmp(token, "-") == 0) {
        result = num1 - num2;
      } else if (strcmp(token, "*") == 0) {
        result = num1 * num2;
      } else if (strcmp(token, "/") == 0) {
        result = num1 / num2;
      } else if (strcmp(token, "^") == 0) {
        result = pow(num1, num2);
      }

      operand[--j] = result;
    }

    dbool = 1; // Reset to 1
    token = strtok(NULL, s);
  }

  printf("%d\n", result);
  return 0;
}

File: Cpp Course Assignments/1 - RPN Calculator in C, Bash, and Python/rpnc.py
###############################################################################
# Annie Shen (ashen7)
# CMPS 109 Harrison - Spring 2018
# Lab 1 - Python3 Reverse Polish Notation (RPN) Calculator
# Due Apr 8th, 2018
###############################################################################
#
# Accepts an arbitrary number of arguments in RPN, evaluates them and prints
# the answer to stdout.

import sys
import shlex
# shlex from: https://stackoverflow.com/questions/899276/python-how-to-parse-strings-to-look-like-sys-argv
test = shlex.split(sys.argv[1])
stack = []

# enumerate from: https://stackoverflow.com/questions/522563/accessing-the-index-in-python-for-loops
for idx, oper in enumerate(test):
    oper = str(oper)

    if oper.isdigit():
        stack.append(oper)

    else:
        # Once reach an operator, pop the newest two numbers on stack, perform
        # the operation with the given operator, and push result back onto stack.
        num2 = int(stack.pop())
        num1 = int(stack.pop())

        if oper is '+':
            result = num1 + num2
        elif oper is '-':
            result = num1 - num2
        elif oper is '*':
            result = num1 * num2
        elif oper is '/':
            result = num1/num2
        elif oper is '^':
            result = num1**num2

        stack.append(result)

# End of operations, return the result, which should be the only item in stack.
print(stack[0])
